### PR TITLE
Resolved Conflict in backend Pagination

### PR DESCRIPTION
**Pull request by Kashish0423**
Hello @Manvityagi  , I had conflicts in the previous PR, i was facing some issues to solve them and wasn't able to identify how to get out of it, so I made a new branch just to reduce the time of getting the work done , please have a look at it and let me know if  there are any changes that should be made .

### CODE REVIEW

File: managers/opportunity/index.js [hunk 1]
```diff
@@ -41,20 +41,139 @@ class opportunityManager {
     }
   }
 
+  // http://localhost:3030/opportunity?limit=4&page=2, Example of how to use the Get Method
+
+  //  page => Means the Current Page that the User Wants
+  //  limit => The count of pages that should be rendered starting from the given page
```
Comment (Manvityagi) [inline managers/opportunity/index.js]: limit - The count of maximum opportunities returned on 1 page. 

The comments should be proper so that no one is misled by them.

File: managers/opportunity/index.js [hunk 2]
```diff
@@ -41,20 +41,139 @@ class opportunityManager {
...
+      console.log('Values of QueryString', queryObject);
+
+      // Here page means which page to start from and limit specifies the number of documnents to be
+      // rendered starting from this page
+
+      let { limit, page } = queryObject;
+
+      // The case when both page and limit are not specified we assign both of them default values
+      if (page == undefined && limit == undefined) {
+        limit = 1;
```
Comment (Manvityagi) [inline managers/opportunity/index.js]: The default limit 1 is very less, by default let's keep it 10

File: managers/opportunity/index.js [hunk 3]
```diff
@@ -41,20 +41,139 @@ class opportunityManager {
...
+        page = 1;
+      }
+
+      //Only limit is 'undefined or 0' , so it is assigned a default value of 1
+      if (limit == undefined) {
+        limit = 1;
```
Comment (Manvityagi) [inline managers/opportunity/index.js]: limit = 10

File: managers/opportunity/index.js [hunk 4]
```diff
@@ -41,20 +41,139 @@ class opportunityManager {
...
+      }
+
+      // The case when starting page is not mentioned or it is Put in as 0 , we use default value
+      if (page == undefined) {
```
Comment (Manvityagi) [inline managers/opportunity/index.js:76]: In comment, you have written that it covers the case of page = 0  also, but the `if` condition covers only `undefined`

File: managers/opportunity/index.js [hunk 5]
```diff
@@ -41,20 +41,139 @@ class opportunityManager {
...
+      // Number of Total Documents Saved in a variable
+      let numDocs = await this.opportunity.countDocuments().exec();
+
+      // TotalPages denote the total number of pages that are possible with the given limit and Documents available
+      let totalPages = Math.floor(numDocs / limit) + (numDocs % limit);
+
+      // Begin here shows the first index of next Page and helps in determining wether or not it is possible to go on the next page
+      let begin = page * limit;
+
+      // let docs= await this.opportunity.find();
```
Comment (Manvityagi) [inline managers/opportunity/index.js]: Remove extra comment

File: managers/opportunity/index.js [hunk 6]
```diff
@@ -41,20 +41,139 @@ class opportunityManager {
...
+
+      // console.log(docs) ;
```
Comment (Manvityagi) [inline managers/opportunity/index.js]: Remove extra comment

File: managers/opportunity/index.js [hunk 7]
```diff
@@ -41,20 +41,139 @@ class opportunityManager {
...
+
+      // If any of the paramter becomes neagtive or Zero
+
+      if (page <= 0 || limit <= 0) {
+        result.results = 'The Parameters cannot be Negative or Zero';
+        return result;
+      }
+
+      // If user enters a Page number greater than the number of Documents present in the Database
+
```
Comment (Manvityagi) [inline managers/opportunity/index.js]: The comments and the code should not have a blank line in between, At some places, there is a blank line after comment, at other places it is not. 

Remove it from everywhere.

**Review by Manvityagi — CHANGES_REQUESTED** [review]
Amazing code other than these little changes, Ready to be merged, just resolve these little things quickly @Kashish0423 

Also, the tests and documentation have not been added for the changes. 
Tests are in the `tests` folder at root of the project, 
and Swagger JSDocs are in `routes/opportunity.js`, 

As a rule of thumb, tests and Documentations should always be updated while making a PR. 
Please do that also.

File: managers/opportunity/index.js (hunk 1 above)
Comment (Kashish0423) [reply managers/opportunity/index.js]: oh yess, my mistake I left this out from the previous code , will correct it

File: managers/opportunity/index.js (hunk 2 above)
Comment (Kashish0423) [reply managers/opportunity/index.js]: okay

File: managers/opportunity/index.js (hunk 4 above)
Comment (Kashish0423) [reply managers/opportunity/index.js:76]: yes Mam, it does if you look at line no 114 this one 

 if (page <= 0 || limit <= 0) {
        result.results = 'The Parameters cannot be Negative or Zero';
        return result;
      }

this part takes care of that case 

and the undefined handles the case if either of page /limit is undefined or both undefined

**Comment by Kashish0423** [comment]
> Amazing code other than these little changes, Ready to be merged, just resolve these little things quickly @Kashish0423
> 
> Also, the tests and documentation have not been added for the changes.
> Tests are in the `tests` folder at root of the project,
> and Swagger JSDocs are in `routes/opportunity.js`,
> 
> As a rule of thumb, tests and Documentations should always be updated while making a PR.
> Please do that also.


Hello mam 
1. In the Tests folders , I can see 3 folders managers, controllers and services  , so where do I need to add test cases , also the Test cases that I am thinking about are like the cases that I try and test the code eg limit=2 and page=3 and something like that but in the file I see some code like using functions (expect() , stub(), spy() )etc for the very first time so I don't know at the moment how to solve it , if possible can you assign it as another issue as I would have to go in to the concepts of the code  in order to implement it and the issue will  remain pending for more time  .

2. As far documentation is concerned , do I have to change it in the apis.md file as well ?

**Comment by Manvityagi** [comment]
> > Amazing code other than these little changes, Ready to be merged, just resolve these little things quickly @Kashish0423
> > Also, the tests and documentation have not been added for the changes.
> > Tests are in the `tests` folder at root of the project,
> > and Swagger JSDocs are in `routes/opportunity.js`,
> > As a rule of thumb, tests and Documentations should always be updated while making a PR.
> > Please do that also.
> 
> Hello mam
> 
> 1. In the Tests folders , I can see 3 folders managers, controllers and services  , so where do I need to add test cases , also the Test cases that I am thinking about are like the cases that I try and test the code eg limit=2 and page=3 and something like that but in the file I see some code like using functions (expect() , stub(), spy() )etc for the very first time so I don't know at the moment how to solve it , if possible can you assign it as another issue as I would have to go in to the concepts of the code  in order to implement it and the issue will  remain pending for more time  .
> 2. As far documentation is concerned , do I have to change it in the apis.md file as well ?

1. Ok
2. No the documentation of the APIs is in `routes/opportunity.js`, open that file, (pull the latest code if you can't see the documentation there), you have to edit the `get API` Documentation there. It's fairly simple, you will understand it by looking at the already existing documentation.

**Comment by Kashish0423** [comment]
Hello @Manvityagi mam please check my latest commit and as of now the branch doesn't have any conflicts and it also passed all the checks , I have made all the necessary changes .

File: routes/opportunity.js [hunk 8]
```diff
@@ -19,8 +19,8 @@ const opportunityManager = new OpportunityManager(),
  *     summary: Retrieve a list of opportunities.
  *     description: Retrieve a list of opportunities.
  *     parameters:
- *      - in: query
- *        name: type
+ *        in: query
+ *        name: page(specifies the Page Number ) and limit (number of results per page )
```
Comment (Manvityagi) [inline routes/opportunity.js]: Why did you remove `type`? 
Please change it back to what it was. 

What you have to do is add on top of this
``` parameters:
       - in: query
         name: type
         schema:
           type: string
           enum:
              - HACKATHON
              - JOB
              - INTERNSHIP
              - SCHOLARSHIP
              - CONFERENCE
              - CODINGCOMPETITION
         description: Retrieve a list of opportunities based on particular type.
  ` 
 
After this before responses, add page and limit 
`     - in: query
        name: page
        schema:
          type: Integer
          default: 1
        description: The Current Page for which the data User Wants
      - in: query
        name: limit
        schema:
          type: Integer
          default: 10
        description: The count of items/documents that should be returned on each page
 ```
 
 Take care of indentation, otherwise it will not work, 
Test your code locally before updating the PR, 
If everything goes the way expected, 
Visiting `http://localhost:3030/playground/` 
would get you something like this(PFA ss) in GET API
image

**Review by Manvityagi — CHANGES_REQUESTED** [review]
Do the requested update, I will merge after that.

File: routes/opportunity.js (hunk 8 above)
Comment (Kashish0423) [reply routes/opportunity.js]: okay mam , on it

Comment (Kashish0423) [reply routes/opportunity.js]: Yes , check the PR , the only difference that i see is the name of First index in your screenshot it is page(specifies...)  but in my case I have  kept the name  as type itself as requested . Rest all same as the above screenshot in my latest commit .

**Review by Manvityagi — APPROVED** [review]

File: routes/opportunity.js (hunk 8 above)
Comment (Manvityagi) [reply routes/opportunity.js]: Great effort and work in PR @Kashish0423 , Merging

**Comment by Manvityagi** [comment]
Fixed #44
